(mini.indentscope) Add option for a prefix symbol

mivort opened the pull request:
This small change adds an ability to use a user-defined symbol for a prefix instead of only a hard-coded space character. The idea is to allow the usage of this plugin with a leading space 'listchar' option enabled.

- [x] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [x] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

## Context

Hi! Thanks a lot for a very cool and minimalistic plugin to provide a nice indent visualization without introducing a performance penalty!

I've been using a `listchars` vim option for a while to visualize leading tabs and whitespaces, so when I've stared using `mini.indentscope` plugin, I've noticed that it overwrites existing `NonText` marks with a blank space:

<img width="593" alt="before" src="https://user-images.githubusercontent.com/5783792/201978241-e583531e-975d-4e54-ab46-6a92ed5aaea4.png">

I think the option to provide a user-defined prefix character instead of hard-coded `' '` could be useful to `listchars` users without introducing a performance penalty. With this patch it's possible to keep the visible whitespaces while getting a highlighted scope under cursor (using the following config):
```lua
vim.o.listchars = 'lead:┆'
require('mini.indentscope').setup { prefix = '┆' }
```

<img width="593" alt="after" src="https://user-images.githubusercontent.com/5783792/201978801-22ac64a9-ef4e-4470-a184-cae47cd54b0c.png">

It can also provide users with ability to customize scope visualization with existing `MiniIndentscopePrefix` highlight.

In any case, thanks again for a great and very useful Neovim plugin!

echasnovski commented:
Hi! Thanks for the PR!

This is an interesting idea. Seems like related to #125.

Would you mind sharing relevant config parts with which you achieved the second screenshot?

mivort commented:
@echasnovski Sure, here's options I've used:
```lua
vim.o.listchars = 'lead:┆'
require('mini.indentscope').setup { prefix = '┆' }
```

mivort commented:
I think I found a way to handle nicely all possible `listchar`s combinations. The idea is to support an 'empty' value, in which case prefix won't be prepended, but rather `virt_text_win_col` increased to the proper amount. That way prefix will be 'transparent' and will not overwrite any listchars.

Here how it looks with `prefix = ''` and `vim.o.listchars = 'leadmultispace:*-'`:

<img width="483" alt="leading" src="https://user-images.githubusercontent.com/5783792/202011214-53f43cd5-7aec-4583-bc2e-52d9d00611c0.png">

I've updated PR and added a remark about empty string value case handling in the documentation.

echasnovski commented:
I am a little bit confused right now... Using `virt_text_win_col` option seems to solve the issue quite nicely even without the new `prefix` option. I thought maybe it was added in Neovim 0.8, but no, it is present in Neovim 0.6.1. And all tests seem to (virtually) pass...

I'll need to investigate this further. Maybe your PR will indirectly lead to removing some code instead of adding one, which is even better :)

mivort commented:
Without at least some option it will make `MiniIndentscopePrefix` redundant and it wouldn't be possible to have something like this (e.g. provide a significantly more visually pronounced indent, which is nice):
``` lua
require('mini.indentscope').setup { symbol = ' ', prefix = ' ' }
vim.cmd [[hi MiniIndentscopeSymbol ctermbg=203 | hi MiniIndentscopePrefix ctermbg=210]]
```
![image](https://user-images.githubusercontent.com/5783792/202147932-b47d02b9-0897-4ef4-8dd1-aa08977daf32.png)

So I think keeping this option may provide a benefit (and also allows to keep the old behavior, since the default value is `' '`).

echasnovski commented:
> Without at least some option it will make `MiniIndentscopePrefix` redundant and it wouldn't be possible to have something like this (e.g. provide a significantly more visually pronounced indent, which is nice):

True. But the only reason for `MiniIndentscopePrefix` was to create "invisible" highlight group to mitigate the (what I thought was) a necessary approach with prefix. At the moment I am leaning towards deprecating this highlight group, but I'll think it through.

echasnovski commented:
In the end, I decided to make a breaking change of removing `MiniIndentscopePrefix` highlight group in favor of always showing single line.

Thanks for making PR and showing `virt_text_win_col` capabilities!
